src/basic/types.h

```diff
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/types.h>
+
+struct stat;
+struct iovec;
+struct siphash;
+
+typedef void (*free_func_t)(void *p);
+typedef void* (*mfree_func_t)(void *p);
+
+typedef void (*hash_func_t)(const void *p, struct siphash *state);
+typedef int (*compare_func_t)(const void *a, const void *b);
+
+struct hash_ops;
+
+typedef void* (*hashmap_destroy_t)(void *p);
+
+/* The base type for all hashmap and set types. Many functions in the implementation take (HashmapBase*)
+ * parameters and are run-time polymorphic, though the API is not meant to be polymorphic (do not call
+ * underscore-prefixed functions directly). */
+typedef struct HashmapBase HashmapBase;
+
+/* Specific hashmap/set types */
+typedef struct Hashmap Hashmap;               /* Maps keys to values */
+typedef struct OrderedHashmap OrderedHashmap; /* Like Hashmap, but also remembers entry insertion order */
+typedef struct Set Set;                       /* Stores just keys */
+typedef struct OrderedSet OrderedSet;         /* Like Set, but also remembers entry insertion order */
+
+typedef struct IteratedCache IteratedCache;   /* Caches the iterated order of one of the above */
+typedef struct Iterator Iterator;
+
+typedef struct Prioq Prioq;
+
+typedef struct FDSet FDSet;
```

Comment:
hmm, this one comes from shared/ and comes with few transitive includes (only 1 with types.h).

Reply:
I think we want the types from src/shared forward declared in here as well to keep things simple so that only one forward declaration header has to be included.